### PR TITLE
perfect restart

### PR DESCRIPTION
**Pull request by ecoon**
It has long been a goal to have a so-called "perfect" restart, where, to machine precision, we are able to continue exactly as we left off a run.

This PR is the start of that.

Effectively it:
1. reverses the order of dependencies, making state not need operators
2. Adds the ability to store Epetra_Vector in State directly, with no special "factory", only a Epetra_Map object as the factory.
3. Adds the ability to store TreeVectors in state.
4. Pass a state object into the SolutionHistory construction, then stash all vectors for solution history (and keep them up-to-date despite many "swap" operations on pointers within the history, e.g. the nth vector is often moved up to the nth+1 vector, and we want to only move the pointer, but when we restart, we have to know it was in the nth+1 position).

Note this is a WIP because there is no test and I haven't actually tried to use it yet.  It just compiles for now, and all existing tests are being checked.  We'll have to add another test to see if this actually works.

Note that the vectors need to be named, so the BDF TI needs to name them.  I will likely update this API a bit so that the TI gets a name (likely given by the PK to be the PK's name) to use with these vectors, thereby enforcing the requirement that they are unique across all BDF_TI instances.  For now, ATS PKs all name their TI through a VerboseObject name, so I'm using that name.  Amanzi will have to think about how it wants to do this, since I suspect that @lipnikov  likes having the hard-coded TI::BDF as the VerboseObject header for ALL of his TIs?

### CODE REVIEW

**Comment by ecoon** [comment]
Ok, still no formal test for this, but I have confirmed with ATS that I now get, to machine precision, the same answer 40+steps into a continued run as I got in the original run.

To do this, I had to move both the SolutionHistory and the TimestepController parameters (e.g. the state associated with number of failed steps, etc, that are used to adjust the adaptive timestep) into State.

This would not yet work for TimestepControllerAdaptive, which stores vectors as well.  Are those vectors the same as those vectors stored in SolutionHistory?  Or do these need to separately be stashed in state as well?

**Comment by lipnikov** [comment]
@ecoon A suggestion made during previous discussion was to use verbose object name and/or its special parameter value. If no object was specified in plist, use plist name. I would be fine with this.

File: src/state/data/Data.hh [hunk 1]
```diff
@@ -211,6 +213,7 @@ template <typename T> Data data(const Teuchos::RCP<T>& p) {
   return Data(std::unique_ptr<Data_Intf>(new Data_Impl<T>(p)));
 }
```
Comment (lipnikov) [inline src/state/data/Data.hh:215]: Empty line at the end is probably not needed.

File: src/state/data/Data_Helpers.cc [hunk 2]
```diff
@@ -352,6 +359,143 @@ bool Initialize<CompositeVector>(
...
+void WriteVis<Epetra_Vector>(const Visualization& vis, const Key& fieldname,
+                             const std::vector<std::string>* subfieldnames,
+                             const Epetra_Vector& vec)
+{
+  vis.Write(fieldname, vec);
+}
+
+template <>
+void WriteCheckpoint<Epetra_Vector>(const Checkpoint& chkp,
+                                      const Key& fieldname,
```
Comment (lipnikov) [inline src/state/data/Data_Helpers.cc:376]: Alignment is off.

File: src/state/evaluators/EvaluatorIndependentFromFile.hh [hunk 3]
```diff
@@ -22,6 +22,8 @@
 
 namespace Amanzi {
 
+class Function;
```
Comment (lipnikov) [inline src/state/evaluators/EvaluatorIndependentFromFile.hh:25]: Class is not mentioned in this header. Remove?

File: src/state/test/state_container.cc [hunk 4]
```diff
@@ -62,8 +61,45 @@ TEST(STATE_FACTORIES_PERSIST) {
...
   s.Setup();
+  Teuchos::RCP<CompositeVector> cv = s.GetPtrW<CompositeVector>("my_vec", Tags::DEFAULT, "my_vec_owner");
+  // putting TreeVectors into state is a little tricky because they are
+  // typically created from existing CompositeVectors.  We setup first, then
+  // Require and stuff the pointer in.
+  s.Require<TreeVector, TreeVectorSpace>("my_tree_vec", Tags::DEFAULT, "my_tree_vec");
+  Teuchos::RCP<TreeVector> tv = Teuchos::rcp(new TreeVector());
+  tv->SetData(cv);
+  s.SetPtr<TreeVector>("my_tree_vec", Tags::DEFAULT, "my_tree_vec", tv);
+
```
Comment (lipnikov) [inline src/state/test/state_container.cc:72]: Too many empty lines

File: src/time_integration/TimestepControllerSmarter.cc [hunk 5]
```diff
@@ -13,15 +13,57 @@
...
+    last_fail_ = S_->GetPtrW<int>(varname, Tags::DEFAULT, name_);
+
+    varname = name_+"_growth_wait_after_fail";
+    S_->Require<int>(varname, Tags::DEFAULT, name_);
+    S_->SetPtr<int>(varname, Tags::DEFAULT, name_, Teuchos::rcp(new int(0)));
+    S_->GetRecordW(varname, Tags::DEFAULT, name_).set_initialized();
+    S_->GetRecordW(varname, Tags::DEFAULT, name_).set_io_checkpoint();
+    growth_wait_after_fail_ = S_->GetPtrW<int>(varname, Tags::DEFAULT, name_);
+  } else {
+    increase_factor_ = Teuchos::rcp(new double(0));
```
Comment (lipnikov) [inline src/time_integration/TimestepControllerSmarter.cc:61]: Converting double/int to an RCP pointer sounds too heavy to me.

**Comment by lipnikov** [comment]
@ecoon Adaptive time step controller stores simple approximations of solution time derivatives. Should be the same as xdot,
I guess. Perhaps it was not done immediately, b/c of missing interfaces to internal state of a time integrator. That time a request for new interfaces could take significant time. I can potentially re-test Darcy flow with controller->state->xdot after this PR is done.

**Comment by ecoon** [comment]
> @ecoon A suggestion made during previous discussion was to use verbose object name and/or its special parameter value. If no object was specified in plist, use plist name. I would be fine with this.

Great, this is what I have in here now then.

File: src/time_integration/TimestepControllerSmarter.cc (hunk 5 above)
Comment (ecoon) [reply src/time_integration/TimestepControllerSmarter.cc:61]: There really isn't another way to do it though.  All data in state is stored in shared pointers (RCPs), and if we want to have local copy of that value and not have to constantly be copying these ints between state and the local version, then we have to store them in shared pointers.

This is still in control code -- it isn't performance critical in anyway.  Sure, it's a bit ugly to constantly be dereferencing these things, but I think it is less ugly than the alternative.  I mean, which is worse:

Current code:
```increase_factor_ = std::min(increase_factor0_, max_increase_factor_);```

Option 1:
```*increase_factor_ = std::min(increase_factor0_, max_increase_factor_);```

Option 2:
```
increase_factor_ = std::min(increase_factor0_, max_increase_factor_);
S_->Assign(name_+"_increase_factor", Tags::DEFAULT, name_, increase_factor_);
```

File: src/state/evaluators/EvaluatorIndependentFromFile.hh (hunk 3 above)
Comment (ecoon) [reply src/state/evaluators/EvaluatorIndependentFromFile.hh:25]: Used at line 71 of this file

**Pull request merged**